ФЗ 11.

NurievLinar opened the pull request.

NurievLinar committed:
ФЗ 11.

NurievLinar commented:
Здравствуйте. перед отправкой задания, прогнал на Postman - всё работает, без ошибок.
На GitHub при проверке зависает на "Run /tmp/wait-for-it.sh -t 60 localhost:8080"

NurievLinar committed:
ФЗ 11.
Удалил неиспользуемые классы:
InMemoryFilmStorage.java
InMemoryUserStorage.java

NurievLinar commented:
Вот ссылка на скриншот Postman.
https://disk.yandex.ru/i/8ReYwCgmlS4tMQ

NurievLinar committed:
ФЗ 11.
редактировал pom.xml

NurievLinar committed:
ФЗ 11.
редактировал pom.xml

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/FilmController.java:
```diff
@@ -34,47 +37,36 @@ public Film addFilm(@Valid @RequestBody Film film){
         log.info("Ставим лайк фильму.");
         filmService.addLikeFilm(id, userId);
-        return filmService.getFilmById(id);
+        //return filmService.getById(id);
```
Не оставляйте в проекте закомментированного кода.

NurievLinar replied:
Исправил.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controllers/GenreController.java:
```diff
@@ -0,0 +1,38 @@
+    private final GenreService genreService;
+
+    @Autowired
+    public GenreController(GenreService genreService) {
```
Не обязательно писать конструктор руками, можно пользоваться аннотацией Lombok @AllArgsConstructor. Аналогично и для других классов.

NurievLinar replied:
Исправил для всех классов.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,59 +1,81 @@
 
     public User getUserById(Integer id) {
-        return userStorage.getById(id);
+        if (id < 1000 && id > 0) {
```
Ограничение с 1000 непонятно. Если оно нужно, то лучше вынести 1000 в константу и дать ей информативное название.

NurievLinar replied:
Исправил.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,58 +1,105 @@
-    public Film getFilmById(Integer id) {
-        return filmStorage.getById(id);
+    public Film getById(Integer id) {
+        if (id < 9900) {
```
Ограничение с 9900 непонятно. Если оно нужно, то лучше вынести 9900 в константу и дать ей информативное название.

NurievLinar replied:
Исправил.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,58 +1,105 @@
+        List<Genre> actualGenreFilm = new ArrayList<>();
+        for (Genre genre : film.getGenres()) {
+            actualGenreFilm.add(genreService.getById(genre.getId()));
+            if (!filmDbStorage.setGenre(film.getId(), genre.getId())) {
```
Не очень хорошо делать запросы к БД в цикле. Это большая нагрузка на БД. Затратно по времени из-за передачи по сети.

NurievLinar replied:
Принято, спасибо. Можно в этом проекте оставлю так? Учту на будущее.

sergeymamonov replied:
Нет, так оставлять нельзя. Запросы к БД далее будут постоянно, и если не разобраться с ними, то дальше будет сложнее. Поэтому рекомендую в случае сложностей обратиться к наставнику за консультацией и разобраться в этом вопросе.

NurievLinar replied:
Переделал.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,58 +1,105 @@
+            if (!actualGenreFilm.contains(genreService.getById(genre.getId()))) {
+                actualGenreFilm.add(genreService.getById(genre.getId()));
+            }
+            if (!filmDbStorage.setGenre(film.getId(), genre.getId())) {
```
Запросы к БД в цикле.

NurievLinar replied:
Исправил.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,58 +1,105 @@
+        List<Genre> currentGenreFilm = genreService.getGenresId(film.getId());
+        for (Genre current : currentGenreFilm) {
+            if (!actualGenreFilm.contains(current)) {
+                filmDbStorage.deleteGenre(film.getId(), current.getId());
```
Запросы к БД в цикле.

NurievLinar replied:
Исправил.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,58 +1,105 @@
     }
 
-
+    private void existsUser(Integer userId) {
```
Методы, которые вызываются только внутри класса, лучше делать приватными.

NurievLinar replied:
Исправлено.

sergeymamonov commented:
> Здравствуйте. перед отправкой задания, прогнал на Postman - всё работает, без ошибок. На GitHub при проверке зависает на "Run /tmp/wait-for-it.sh -t 60 localhost:8080"

Добрый день. Попросил вашего наставника подключиться к этому вопросу.

NurievLinar committed:
ФЗ 11.
редактировал pom.xml

NurievLinar committed:
ФЗ 11.
исправления

NurievLinar committed:
ФЗ 11.
data.sql, merge into вместо insert into

NurievLinar committed:
Merge remote-tracking branch 'origin/add-database' into add-database

NurievLinar commented:
Ошибка с проверкой в GitHub решилась путём изменения в data.sql `insert into` на `merge into`. Помог наставник.

NurievLinar committed:
ФЗ 11.
Добавил новую функциональность - сохранение состояния данных в БД.
Исправил замечания.
Устранил ошибку при проверке на GitHub.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -2,58 +2,77 @@
 public class UserService {
 
-    UserStorage userStorage;
+    private static final int LIMIT = 1000;
```
Из названия константы непонятно, чего именно лимит.

NurievLinar replied:
Исправил.

NurievLinar committed:
ФЗ 11.
Добавил новую функциональность - сохранение состояния данных в БД.
Исправил замечания.
Устранил ошибку при проверке на GitHub.
Исправлено:
- Исключил запросы к БД в цикле.
- Убрал константы.

NurievLinar committed:
ФЗ 11.
Добавил новую функциональность - сохранение состояния данных в БД.
Исправил замечания.
Устранил ошибку при проверке на GitHub.
Исправлено:
- Исключил запросы к БД в цикле.
- Убрал константы.
- Исправлены тесты.

NurievLinar committed:
ФЗ 11.
Добавил новую функциональность - сохранение состояния данных в БД.
Исправил замечания.
Устранил ошибку при проверке на GitHub.
Исправлено:
- Исключил запросы к БД в цикле.
- Убрал константы.
- Исправлены тесты.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -2,58 +2,71 @@
-            if (userStorage.getById(idFriend).getFriends().contains(idFriendUser)) {
-                commonFriend.add(userStorage.getById(idFriendUser));
-            }
+        Set<Integer> common = new HashSet<>(userDbStorage.findAllFriends(idUser));
```
Разве здесь findAllFriends возвращает массив с дубликатами? Предложил бы посмотреть в дебагере. Но даже если возвращает, то не самое лучше решение через множество делать. Лучше и быстрее будет работать, если сделать новый запрос, который будет как в методе findAllFriends, только с SQL словом DISTINCT. Общее правило такое, всю фильтрацию, которую можно сделать на стороне БД, лучше сделать на стороне БД. БД это сделает быстрее, так как в БД используются индексы и специальные структуры данных. И после фильтрации в БД ответ будет меньше, значит быстрее будет передан по сети.

NurievLinar replied:
Исправлено

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -2,58 +2,71 @@
-        for (Integer friendId : userStorage.getById(idUser).getFriends()) {
-            friends.add(userStorage.getById(friendId));
+        for (Integer friendId : idFriends) {
+            friends.add(getUserById(friendId));
```
От того, что запрос в БД обернут в метод, их не стало меньше) Запросы к БД как были в цикле, так они и остаются.

NurievLinar replied:
Исправлено

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/storage/GenreDbStorage.java:
```diff
@@ -0,0 +1,62 @@
+        List<Integer> idGenres = jdbcTemplate.queryForList(sqlQuery, Integer.class);
+        List<Genre> genres = new ArrayList<>();
+        for (Integer id : idGenres) {
+            genres.add(new Genre(id, getById(id)));
```
Запросов к БД меньше не стало. Цель именно сократить запросы к БД.

NurievLinar replied:
Исправлено

sergeymamonov reviewed (changes requested):
Описание коммитов должно быть подробно. Не должно быть несколько коммитов с одинаковым названием. В коллективной разработке, ваши коллеги будут видеть ваши коммиты, вам может потребоваться откатиться на какой-то коммит. Если они все называются одинаково, то будет сложно понять, на какой откатываться.

Запросы к БД в цикле остались. Подсветил пару мест.

NurievLinar committed:
ФЗ 11.
UserDbStorage - Исправлено обращение к БД в цикле.

NurievLinar committed:
ФЗ 11.
GenreDbStorage - Исправлено обращение к БД в цикле.

NurievLinar committed:
ФЗ 11.
FilmTests - Исправлены названия методов в тестах.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -2,57 +2,72 @@
-        return filmStorage.getFilms();
+        var result = filmDbStorage.getFilms();
+        for (Film film : result) {
+            film.setGenres(genreService.getGenresByFilmId(film.getId()));
```
В getGenresByFilmId выполняется запрос к БД. Значит в цикле мы делаем запросы к БД.

NurievLinar replied:
Исправлено.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,224 @@
+    public List<Film> mostPopulars(Integer limit) {
+        log.info("FilmDbStorage. mostPopulars. limit: {}", limit);
+        List<Film> allFilms = getFilms();
+        for (Film film : allFilms) {
```
В этом цикле есть запросы к БД.

NurievLinar replied:
Исправлено.

NurievLinar committed:
ФЗ 11.
FilmDbStorage. List<Film> mostPopulars - В методе исправлен цикличный запрос к БД.
FilmService. List<Film> getFilms() - Исправлен цикличный запрос к БД.

NurievLinar commented:
От чистого сердца Благодарю! Хорошо меня натаскали!)
